src/server/campaignd/addon_utils.cpp

```diff
@@ -152,4 +153,74 @@ void add_license(config& cfg)
 	copying["contents"] = contents;
 }
 
+static const std::string file_hash(const config& file)
+{
+	std::string hash = file["hash"].str();
+	if(hash.empty()) {
+		hash = utils::md5(file["contents"].str()).base64_digest();
+	}
+	return hash;
+}
+
+static bool comp_file_hash(const config& file_a, const config& file_b)
+{
+	return file_hash(file_a) == file_hash(file_b);
+}
+
+//! Surround with [dir][/dir]
+static void write_difference(config& pack, const config& from, const config& to, bool with_content)
+{
+	pack["name"] = to["name"];
+
+	for(const config& f : to.child_range("file")) {
+		bool found = false;
+		for(const config& d : from.child_range("file")) {
+			found |= comp_file_hash(f, d);
+			if(found)
+				break;
+		}
+		if(!found) {
+			config& file = pack.add_child("file");
+			file["name"] = f["name"];
+			if(with_content) {
+				file["contents"] = f["contents"];
+				file["hash"] = file_hash(f);
+			}
+		}
+	}
+
+	for(const config& d : to.child_range("dir")) {
+		const config& origin_dir = from.find_child("dir", "name", d["name"]);
+		if(origin_dir) {
+			config& dir = pack.add_child("dir");
+			write_difference(dir, origin_dir, d, with_content);
+		}
```

Comment:
this looks liek it'll ignore newly added directories.

Reply:
Will probably fix it by feeding an empty dir if the directory is not found.

Reply:
Done.

Reply:
I'm wondering, instead of calling write_difference again, wouldn't it have teh same effect to just add the whole of d? `pack.add_child("dir", d);` (instead of calling write_difference)


Also It seems like this cannot delete directories: if the funciton is called with with_content=no then, it will detect that all files here are 'missing' but it probably  won't be able to delete the directory.

Reply:
The first point is worth considering, indeed.

The list of files to be removed is written in a `removelist` with `with_content=no`.  If the directory after purging the files is empty, it gets removed:
https://github.com/wesnoth/wesnoth/blob/3764ac90322a883a2d6910a89707744100c26c6b/src/addon/manager.cpp#L318-L320

Reply:
>  If the directory after purging the files is empty, it gets removed:

hmm so it impossible to upload addons with empty directories ?

Reply:
They may be possible in a full pack, but if a directory was containing files, but as a result of an update pack it got empty (the purge check goes across the update pack removelist), it will be deleted.

Reply:
Also, the proposal to just append the whole directory if it's new, won't work correctly in the case of `with_content=no` when we need to make a list of files without content.

Reply:
> They may be possible in a full pack, but if a directory was containing files and as a result of an update pack it got empty (the purge check goes across the update pack removelist), it will be deleted

So technicially the addon manager might delete empty directories unexpectedly, i don't have a problem with this since i don't see a use-case for empty directories, but its definitely a change that should be documented.

Reply:
Fine, I can something about it on the wiki after the changes are out.

Reply:
There's nowhere in the wiki where people who need to know about this (who??) would see it, really. It's not important anyway.

If this involves `#ifhave` or the equivalent Lua thing, perhaps just drop a note somewhere in there telling people not to rely on the existence of empty directories.

Reply:
If they're using git for version control, empty directories aren't saved either, so that should reduce the incidence of the problem.

Reply:
`filesystem::dir_size(dir)` ought to (as in, I'm looking at it now and think it already does) return 0 rather than -1 for an empty directory.

If you already have a setup for testing this, please would you? It would take a bit of setting up to test myself, as `purge_addon` and `purge_dir` are only called when upgrading an add-on with removed files; they aren't called using the "Remove add-ons" dialog, nor when deleting an add-on via the add-on manager's `-` button.

Reply:
What needs to be tested? Just that dir_size returns 0 on an empty directory?

Reply:
Good point, I can call it from somewhere else to test.

* It returns 0 for an empty directory
* It throws `boost::filesystem::filesystem_error` for a path that points to a nonexistent directory

The behavior of purge_addon and purge_dir also need to be tested, because it's definitely not expecting either of the above results. The first issue is that it probably leaves empty directories (which is trivial), however errors deleting a file are likely to leave the add-on in a partially-installed state (admittedly it probably already wasn't in the right state anyway).